### PR TITLE
feat: use custom X-Ignored-Error-Code header with Apollo queries

### PR DESCRIPTION
**Pull request by nikomakela**
An alternative to #457 .

HH-325.

### Apollo Router
Change the router configuration: Propagate all headers to the Events-proxy.

### Clients
Set `X-Ignored-Error-Code: 'UNPOPULATED_MANDATORY_DATA'` to query by using a Apollo-client middleware. It does not matter if the header is in every query, but it's mostly needed with the event list queries and at this point it's only handled by the events-graphql-proxy.

Relates to #456.

----

### Test instrocutions

It's easiest to use the local router. The errenous article that is good for testing is in the production CMS.

```
    "docker:graphql-router:events:serve": "cross-env FEDERATION_CMS_ROUTING_URL=https://tapahtumat.content.api.hel.fi/graphql FEDERATION_EVENTS_ROUTING_URL=http://docker.for.mac.localhost:4100/proxy/graphql FEDERATION_UNIFIED_SEARCH_ROUTING_URL=https://kuva-unified-search.api.test.hel.ninja/search FEDERATION_VENUES_ROUTING_URL=https://venue-graphql-proxy.test.hel.ninja/proxy/graphql docker-compose -f docker-compose.router.yml up" 
```
Changing the .env.local might be needed to allow unauthorized requests: `NEXT_PUBLIC_ALLOW_UNAUTHORIZED_REQUESTS=0`

Then navigate to http://localhost:3000/fi/artikkelit/musiikki/iloa-ja-valoa-kulttuurikeskuksista

You can configure the used headers from the `AppConfig`.

Test result:

When the new header is used, the router is propagating all the headers to events-proxy and the event-proxy has the #456 features on, the errenous page is shown

image

image

### CODE REVIEW

File: packages/components/src/apollo/EventsFederationApolloClient.ts [hunk 1]
```diff
@@ -43,6 +43,14 @@ type GraphQLErrorsExtensionResponse = {
   status: number;
 };
 
+/**
+ * The Events-graphql-proxy offers a way to ignore some custom ignorable graphql errors.
+ * The 'X-Ignored-Error-Code' header is used to list the errors wanted to be ignored.
+ * */
+export const ignoredErrorCodesHeader = {
+  'X-Ignored-Error-Code': 'UNPOPULATED_MANDATORY_DATA',
+};
```
Comment (nikomakela) [inline packages/components/src/apollo/EventsFederationApolloClient.ts:52]: These should be moved to some constants file and they should be given by the app through the apollo configuration property from appConfig file.

File: packages/components/src/apollo/EventsFederationApolloClient.ts [hunk 2]
```diff
@@ -159,6 +170,23 @@ class EventsFederationApolloClient {
...
+   * Add custom headers to the requests.
+   * NOTE: If using router as a proxy to a subgraph,
+   * remember to propage all these headers to the subgraph
+   * */
+  getCustomHeadersMiddleware() {
+    return new ApolloLink((operation, forward) => {
+      operation.setContext(({ headers = {} }) => ({
+        headers: {
+          ...headers,
+          ...ignoredErrorCodesHeader,
```
Comment (nikomakela) [inline packages/components/src/apollo/EventsFederationApolloClient.ts:183]: Should be read from apollo-client config propperty which is populated by appConfig. Chech if there are any to deal the nulls

**Review by karisal-anders — APPROVED** [review]

**Comment by terovirtanen** [comment]
**Federation Router branch is deployed to platta: https://events-graphql-federation-events-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Federation Router branch is deployed to platta: https://events-graphql-federation-hobbies-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Federation Router branch is deployed to platta: https://events-graphql-federation-sports-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __failed__ for https://harrastukset-pr458.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __failed__ for https://tapahtumat-pr458.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://liikunta-pr458.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr458.dev.hel.ninja** :rocket::rocket::rocket:

**Pull request merged**